### DISCUSSION TITLE
Default templates have analyzer warnings

### DISCUSSION THREAD

**Discussion by Youssef1313** [dotnet/aspnetcore#23561] [General] [answered]
(I'm not sure whether this should be a discussion or an issue)

1. [Program](https://github.com/dotnet/aspnetcore/blob/master/src/ProjectTemplates/Web.ProjectTemplates/content/EmptyWeb-CSharp/Program.cs) class can be made static.

2. [`ConfigureServices`](https://github.com/dotnet/aspnetcore/blob/master/src/ProjectTemplates/Web.ProjectTemplates/content/EmptyWeb-CSharp/Startup.cs#L17) can be marked static. (Not sure if it should).

    - **Note:** the analyzer suggests that Startup class be marked as static, but actually it can't as it's used as a type argument.

3. Parameter `services` in `ConfigureServices` is never used. However, most likely it will be used later. So, I'm okay if this is ignored if there isn't a good solution to it.

4. Consider calling ConfigureAwait on [awaited task](https://github.com/dotnet/aspnetcore/blob/master/src/ProjectTemplates/Web.ProjectTemplates/content/EmptyWeb-CSharp/Startup.cs#L35). (Might be not very important as it's just hello, world. But I think it's a good practice to add ConfigureAwait(false);

**Comment by Youssef1313** [comment, marked as answer]
cc: @stephentoub for the last point.

**Reply by stephentoub** [reply]
ASP.NET Core templates shouldn't have CA2007 enabled by default; ConfigureAwait is for library code, not app code. Is that something you enabled manually, or was that warning raised by default?

**Reply by Youssef1313** [reply]
Just installed the analyzer package from NuGet after creating a new project.

**Reply by Youssef1313** [reply]
Is there a reason to not add ConfigureAwait(false) in a case like that? Would it be better if ConfigureAwait(true) was added to the template?

**Reply by stephentoub** [reply]
Ok, so not by default, that's good then. Which package?

**Reply by stephentoub** [reply]
> Is there a reason to not add ConfigureAwait(false) in a case like that? Would it be better if ConfigureAwait(true) was added to the template?

https://devblogs.microsoft.com/dotnet/configureawait-faq/

**Reply by Youssef1313** [reply]
> 
> 
> Ok, so not by default, that's good then. Which package?

Microsoft.CodeAnalysis.FxCopAnalyzeres v3.0.0

**Reply by Youssef1313** [reply]
> 
> 
> > Is there a reason to not add ConfigureAwait(false) in a case like that? Would it be better if ConfigureAwait(true) was added to the template?
> 
> https://devblogs.microsoft.com/dotnet/configureawait-faq/

`"if you’re writing app-level code, do not use ConfigureAwait(false)."`, I think this is the conclusion. So from my very basic understanding, it's intentional not to add ConfigureAwait(false). Do you think adding ConfigureAwait(true) to be explicit about this is a good idea?

**Reply by stephentoub** [reply]
> it's intentional not to add ConfigureAwait(false)

Yes

> Do you think adding ConfigureAwait(true) to be explicit about this is a good idea?

No :smile:

**Reply by Youssef1313** [reply]
Thanks. Just few questions I still have.

> > Do you think adding ConfigureAwait(true) to be explicit about this is a good idea?
> 
> No 😄

Is the reason just a matter of style and/or unnecessary change to the default template?

---

Is there something that can be done to make the analyzer package not produce these warnings?

**Reply by stephentoub** [reply]
> Is the reason just a matter of style and/or unnecessary change to the default template?

It's unnecessary clutter for zero gain.

> Is there something that can be done to make the analyzer package not produce these warnings?

Sure, disable the rule.  The FxCopAnalyzers package is primarily about libraries; that's what the "FX" means: framework.

**Reply by Youssef1313** [reply]
@stephentoub, Thanks for the help answering this. I really appreciate that very much.
I'll mark as answered now.
